lib/Dialect/ESI/runtime/cpp/lib/Manifest.cpp

```diff
@@ -0,0 +1,225 @@
+//===----------------------------------------------------------------------===//
+
+static AppID parseID(const nlohmann::json &json) {
+  std::optional<uint32_t> idx;
+  if (json.contains("index"))
+    idx = json.at("index").get<uint32_t>();
+  return AppID{json.at("name").get<std::string>(), idx};
+}
+
+static ModuleInfo parseModuleInfo(const nlohmann::json &mod) {
+  auto getAny = [](const nlohmann::json &json) -> std::any {
+    if (json.is_string())
+      return json.get<std::string>();
+    else if (json.is_number_integer())
+      return json.get<int64_t>();
+    else if (json.is_number_unsigned())
+      return json.get<uint64_t>();
+    else if (json.is_number_float())
+      return json.get<double>();
+    else if (json.is_boolean())
+      return json.get<bool>();
+    else if (json.is_null())
+      return json.get<std::nullptr_t>();
+    else
+      throw std::runtime_error("Unknown type in manifest: " + json.dump(2));
+  };
+  std::map<std::string, std::any> extras;
+  for (auto &extra : mod.items())
+    if (extra.key() != "name" && extra.key() != "summary" &&
+        extra.key() != "version" && extra.key() != "repo" &&
+        extra.key() != "commit_hash" && extra.key() != "symbolRef")
+      extras[extra.key()] = getAny(extra.value());
+
+  auto value = [&](std::string key) -> std::optional<std::string> {
+    auto f = mod.find(key);
+    if (f == mod.end())
+      return std::nullopt;
+    return f.value();
+  };
+  return ModuleInfo{value("name"), value("summary"),     value("version"),
+                    value("repo"), value("commit_hash"), extras};
+}
+
+//===----------------------------------------------------------------------===//
+// ManifestProxy class implementation.
+//===----------------------------------------------------------------------===//
+
+internal::ManifestProxy::ManifestProxy(std::string manifestStr,
+                                       Manifest &manifest) {
+  manifestJson = nlohmann::ordered_json::parse(manifestStr);
+
+  for (auto &mod : manifestJson.at("symbols"))
+    symbolInfoCache.insert(
+        std::make_pair(mod.at("symbolRef"), parseModuleInfo(mod)));
+}
+
+std::unique_ptr<Design>
+internal::ManifestProxy::buildDesign(Accelerator &acc) const {
+  auto designJson = manifestJson.at("design");
+  std::vector<std::unique_ptr<Instance>> children =
+      getChildInstances(acc, designJson);
+  return std::make_unique<Design>(getModInfo(designJson), std::move(children));
+}
+
+std::optional<ModuleInfo>
+internal::ManifestProxy::getModInfo(const nlohmann::json &json) const {
+  auto instOfIter = json.find("inst_of");
+  if (instOfIter == json.end())
+    return std::nullopt;
+  auto f = symbolInfoCache.find(instOfIter.value());
+  if (f != symbolInfoCache.end())
+    return f->second;
+  return std::nullopt;
+}
+
+std::vector<std::unique_ptr<Instance>>
+internal::ManifestProxy::getChildInstances(Accelerator &acc,
+                                           const nlohmann::json &json) const {
```

Comment:
hmm... Why is `acc` passed by non-const reference here? Secondly, `acc` doesn't seem to be used.

Reply:
At some point, I think I'm gonna need it pretty pervasively to hook up communications. But not for now.